src/core/border_router/routing_manager.cpp

```diff
@@ -1632,6 +1651,86 @@ void RoutingManager::DiscoveredPrefixTable::SignalTableChanged(void)
     mSignalTask.Post();
 }
 
+void RoutingManager::DiscoveredPrefixTable::ProcessNeighborAdvertMessage(
+    const Ip6::Nd::NeighborAdvertMessage &aNaMessage,
+    const Ip6::Address &                  aSrcAddress)
+{
+    Router *router = mRouters.FindMatching(aSrcAddress);
+
+    VerifyOrExit(router != nullptr);
+    VerifyOrExit(aNaMessage.IsValid());
+    UpdateRouterOnRx(*router);
+
+exit:
+    return;
+}
+
+void RoutingManager::DiscoveredPrefixTable::UpdateRouterOnRx(Router &aRouter)
+{
+    aRouter.mNsProbeCount = 0;
+    aRouter.mTimeout      = TimerMilli::GetNow() + Random::NonCrypto::AddJitter(Router::kActiveTimout, Router::kJitter);
+
+    mRouterTimer.FireAtIfEarlier(aRouter.mTimeout);
+}
+
+void RoutingManager::DiscoveredPrefixTable::HandleRouterTimer(void)
+{
+    TimeMilli now      = TimerMilli::GetNow();
+    TimeMilli nextTime = now.GetDistantFuture();
+
+    for (Router &router : mRouters)
+    {
+        if (router.mNsProbeCount > Router::kMaxNsProbes)
+        {
+            continue;
+        }
+
+        if (router.mTimeout <= now)
+        {
+            router.mNsProbeCount++;
+
+            if (router.mNsProbeCount > Router::kMaxNsProbes)
+            {
+                LogInfo("Timed out waiting for response to Neighbor Solicitations to %s",
+                        router.mAddress.ToString().AsCString());
+                Get<RoutingManager>().mRsSender.Start();
```

Comment:
Should we only send an RS if there are no more usable prefixes?

Reply:
Do you mean skip sending RS, if the info from this "router" does not impact any of the decisions (i.e., what we decide to publish in net data and/or whether or not to advertise our local on-link prefix)?

I considered something similar initially, i.e. just monitoring and sending NS to routers that are impacting some decision. This became complex and rather tricky since as things change (e.g., we receive from new router, or a prefix is removed/aged, etc), info from this router may become important and impact our decision and then we may rely on stale info (if we don't monitor the router is still present and active).

-----

A related idea/thought: If there are other Thread BRs (connected to same Thread mesh), we receive and process RAs from them, but since all prefixes are from net data or match our on-link prefix, we don't save any of the prefixes which cause the "router" entry to have no prefix and therefore removed from the list. So practically, we should not have any other Thread BRs (connected to same network) in our "router" list and won't monitor and send NS to any of them.